server/protocols/webadmin/webadmin-data/src/test/java/org/apache/james/webadmin/routes/RemoveRegexMappingRoutesTest.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.james.rrt.lib.Mapping;
+import org.apache.james.rrt.lib.MappingSource;
+import org.apache.james.rrt.memory.MemoryRecipientRewriteTable;
+import org.apache.james.webadmin.WebAdminServer;
+import org.apache.james.webadmin.WebAdminUtils;
+import org.eclipse.jetty.http.HttpStatus;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import io.restassured.RestAssured;
+import io.restassured.filter.log.LogDetail;
+import io.restassured.http.ContentType;
+
+class RemoveRegexMappingRoutesTest {
+
+    private WebAdminServer webAdminServer;
+    private MemoryRecipientRewriteTable memoryRecipientRewriteTable;
+
+    @BeforeEach
+    void beforeEach() throws Exception {
+        DNSService dnsService = mock(DNSService.class);
+        DomainList domainList = new MemoryDomainList(dnsService);
+        memoryRecipientRewriteTable = new MemoryRecipientRewriteTable();
+        memoryRecipientRewriteTable.setDomainList(domainList);
+        domainList.addDomain(Domain.of("domain.tld"));
+
+        webAdminServer = WebAdminUtils
+            .createWebAdminServer(new RemoveRegexMappingRoutes(memoryRecipientRewriteTable))
+            .start();
+
+        RestAssured.requestSpecification = WebAdminUtils.buildRequestSpecification(webAdminServer)
+            .setBasePath(RemoveRegexMappingRoutes.BASE_PATH)
+            .log(LogDetail.METHOD)
+            .build();
+    }
+
+    @AfterEach
+    void stop() {
+        webAdminServer.destroy();
+    }
+
+    @Test
+    void removeRegexMappingShouldReturnNoContentWhenSuccess() throws Exception {
+        MappingSource mappingSource = MappingSource.fromUser(User.fromUsername("abc@domain.tld"));
+        memoryRecipientRewriteTable
+            .addRegexMapping(mappingSource, "^[aeiou]");
+
+        with()
+            .delete("abc@domain.tld/targets/^[aeiou]")
+        .then()
+            .statusCode(HttpStatus.NO_CONTENT_204)
+            .contentType(ContentType.JSON);
+
+        assertThat(memoryRecipientRewriteTable.getStoredMappings(mappingSource))
+            .doesNotContain(Mapping.regex("^[aeiou]"));
+    }
+
+    @Test
+    void removeRegexMappingShouldAllowUserWithoutDomain() throws Exception {
+        MappingSource mappingSource = MappingSource.fromUser(User.fromUsername("abcdomaintld"));
+        memoryRecipientRewriteTable
+            .addRegexMapping(mappingSource, "^[aeiou]");
+
+        with()
+            .delete("abcdomaintld/targets/^[aeiou]")
+        .then()
+            .statusCode(HttpStatus.NO_CONTENT_204)
+            .contentType(ContentType.JSON);
+
+        assertThat(memoryRecipientRewriteTable.getStoredMappings(mappingSource))
+            .doesNotContain(Mapping.regex("^[aeiou]"));
+    }
+
+    @Test
+    void removeRegexMappingShouldReturnNotFoundWhenSourceIsEmpty() {
+        with()
+            .delete("/targets/^[aeiou]")
+        .then()
+            .statusCode(HttpStatus.NOT_FOUND_404)
```

Comment:
We tend to like `Idempotent` APIs : applying 2 time the same action will give two time the same result.

Rationals: we care about the entry not being here any longer. We don't care if we are the ones deleting it. Pushing that kind of 404/409 things leads to unecessary complexity and synchronisation. We thus keep things simple.

Here you are not idempotent: You are returning 404 in some cases and 204 in some others.

To be idempotent you need to always return 204, even if the record was not present.

Reply:
IMO the return code 404 for the situation that source or regex is missing or both of them are missing is quite reasonable. Duc explained to me that when one of the parameters, or both of them, were missing, the matching algorithm of Spark will reject the deleting operation, and code 404 would be returned.
204 code for successfully deleted (when source and regex are valid or accepted)

Reply:
You are talking about missing parameters, and yes it is 404 in that case. Benoit is talking to you about valid parameters, but in this case with the information not existing, returning 204 still counting it as a "success".

I know this is a bit difficult to get as a concept, I had that problem too (and maybe still have sometimes), but here is not really about pure logic, but more of a design choice. As Benoit said, the design choice here is being idempotent. Meaning a same request should return always the same result. So in a case of delete, if you want to always to return the same result, you have to accept that even if the resource doesn't exist, or existed previously but not anymore, we still return 204.

Reply:
actually that comment would apply more to the `removeRegexMappingShouldReturnBadRequestWhenRegexIsInvalid` test

Reply:
I got your ideas. New tests were added to guarantee `Idempotent`.